### ISSUE TITLE
Service to Require PRs

### ISSUE THREAD

**Issue by jakirkham** [conda-forge/conda-forge-webservices#75]
Much of conda-forge runs on the notion that PRs from one's fork is a good idea. This is to ensure packages are not released prematurely (before merging), to ensure the linter has a chance to run, and to ensure things actually do build right on the CIs before deploying. Also many future services like providing automated welcome messages for new members ( https://github.com/conda-forge/conda-forge-webservices/issues/73 ) and the re-rendering service ( https://github.com/conda-forge/conda-forge-webservices/issues/74 ) rest on the notion of having PRs or at least performing better when there are PRs.

However, we lack any real mechanism to enforce this behavior ATM. This issue is a write-up of a potential service that simply requires a PR be made. Though maybe we can convince GitHub to take this on long term, this should be simple enough for us to do in the meantime.

The idea behind this is to provide a service that adds an always passing status check to PRs and makes all new branches ( or at least all deployment branches https://github.com/conda-forge/conda-forge.github.io/issues/87 ) require this check. The net effect is that people will be rejected from directly pushing to such branches and will instead need to open a PR to at least get that status check before pushing ahead. Even if they ignore all other status checks, there will at least be a place for them to occur.

As whitelisting of branches is so highly demanded and we lack the ability to make status checks required for all branches, this will likely couple nicely with that whitelisting functionality ( https://github.com/conda-forge/conda-forge.github.io/issues/87 ). The net result being only deployment branches have this constraint. Though some fine tuning will be needed to get the timing worked out.

This could also later could get other options to enable status checks for other CIs maybe based on a config file (e.g. `conda-forge.yml`) or similar. This way feedstock maintainers could require CI status checks if they wish.

cc @asmeurer

**Comment by asmeurer** [comment]
Maybe I'm missing something, but how is this different from just enabling branch protection with required status checks? That disables directly pushing to a branch.

**Comment by jakirkham** [comment]
I thought you still had to pick at least one service for that to really work or no?

**Comment by asmeurer** [comment]
You do. I guess you don't want to require that the builds pass? I would want that for all the feedstocks I maintain, but I think I remember you saying there might be cases where you don't want that. What about the linter?

**Comment by jakirkham** [comment]
> I guess you don't want to require that the builds pass?

Correct.

> I would want that for all the feedstocks I maintain, but I think I remember you saying there might be cases where you don't want that.

So the last point is in particular trying to provide a way for you to address that need in combination with this service since giving everyone admin permissions is not permissible.

> What about the linter?

Thought about that, but the linter can be funny sometimes like with unicode characters or environment variables in jinja. So I'm not ready to tighten that down yet.

Though personally I do try to find linter acceptable solutions if possible.

<hr>
In any event, the hard part is not the service and service status. That turns out to be really easy.

The tricky part is listening to branch creation events and adding these status requirements according to some config.

There is also registering all existing feedstocks with this service. This is doable, but timeouts and API limitation issues could make this a bit tedious. Will have to see.

We probably will have to do this in phases to get this in. Right now I think we should try to hash out what the config would look like for enabling other CIs and then give some thought to the timing issues w.r.t. registering new branches with the service.

**Comment by asmeurer** [comment]
Well I hope you make some way for feedstock maintainers to require all status checks if they want them. You might even consider making that the default (although there is a potential problem there: what if the config is in the repo, and you need to change a certain status that is currently failing to not required. How can you merge the change to the config?).

**Comment by jakirkham** [comment]
Yeah, this seems like a plausible way forward IMHO. Though I don't think we are going to require all status checks by default not only for the reasons that you have mentioned, but because there are cases where we have to accept a failure and merge.

**Comment by xoviat** [comment]
You don't have to require all status checks. You only need to require a status check called "is submitted as a PR," which would essentially always be true for PRs.

**Comment by xoviat** [comment]
Actually you could just reuquire a PR review.
